File: ts/src/parkingBoy/baseParkingBoy.ts
/** @format */

import {ParkingLot} from '../parkingLot/parkingLot'
import {ParkingBoyService} from './parkBoyService'

export abstract class BaseParkingBoy {
    protected constructor(private parkingLots: ParkingLot[], private parkingBoyService: ParkingBoyService) {}

    public get availableSpaces() {
        return this.parkingLots.reduce((result, {availableSpaces}) => result + availableSpaces, 0)
    }

    public get hasAvailableSpaces() {
        return this.parkingLots.some(({hasAvailableSpaces}) => hasAvailableSpaces)
    }

    public findOneValidParkingLot() {
        return this.parkingBoyService.findOneValidParkingLot(this.parkingLots)
    }
}

File: kt/src/main/kotlin/com/github/scglwsj/parking/boy/BaseParkingBoy.kt
package com.github.scglwsj.parking.boy

import com.github.scglwsj.parking.lot.ParkingLot

abstract class BaseParkingBoy protected constructor(
    private val parkingLots: List<ParkingLot>,
    private val parkingBoyService: ParkingBoyService
) {
    val availableSpaces: Int get() = parkingLots.sumBy { it.availableSpaces }
    val hasAvailableSpaces: Boolean get() = parkingLots.any { it.hasAvailableSpaces }
    fun findOneValidParkingLot(): ParkingLot? = parkingBoyService.findOneValidParkingLot(parkingLots)
}
